### PR TITLE
add ParquetBucketMetadata and ParquetTypeFileOperations

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

File: scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/AvroUtils.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.beam.sdk.extensions.smb;
+
+import com.google.common.base.Preconditions;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.beam.sdk.coders.*;
```
Comment (ClaireMcGinty) [inline scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/AvroUtils.java:23]: i don't think you need the wildcard - we're only importing `Coder` itself it looks like

File: scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/ParquetBucketMetadata.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+ */
+package org.apache.beam.sdk.extensions.smb;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.reflect.ReflectData;
+import org.apache.beam.sdk.coders.*;
```
Comment (ClaireMcGinty) [inline scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/ParquetBucketMetadata.java:25]: same comment as above

File: scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/ParquetBucketMetadata.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    switch (getRecordType(recordClass)) {
+      case AVRO:
+        return AvroUtils.validateKeyField(
+            keyField,
+            keyClass,
+            new ReflectData(recordClass.getClassLoader()).getSchema(recordClass));
+      case SCALA:
+        return validateScalaKeyField(keyField, keyClass, recordClass);
+    }
+    throw new IllegalStateException("Should never reach here");
```
Comment (ClaireMcGinty) [inline scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/ParquetBucketMetadata.java:211]: shouldn't this go in the `default` case of the switch statement?

File: scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/ParquetBucketMetadata.java [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+  // Logic for dealing with Avro records vs Scala case classes
+  ////////////////////////////////////////////////////////////////////////////////
+
+  private static RecordType getRecordType(Class<?> recordClass) {
+    if (GenericRecord.class.isAssignableFrom(recordClass)) {
+      return RecordType.AVRO;
+    } else if (scala.Product.class.isAssignableFrom(recordClass)) {
+      return RecordType.SCALA;
+    } else {
+      throw new IllegalArgumentException("Unsupported record class " + recordClass.getName());
```
Comment (ClaireMcGinty) [inline scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/ParquetBucketMetadata.java:197]:
```suggestion
      throw new IllegalArgumentException("Unsupported record class " + recordClass.getName() +". Must be an Avro record or a Scala case class.");
```
added a suggestion to make the error message a little clearer (for example, if someone tries to use a Scala non-case-class, they might be confused by the error msg).

**Review by ClaireMcGinty — APPROVED** [review]

File: scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/ParquetBucketMetadata.java (hunk 3 above)
Comment (nevillelyh) [reply scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/ParquetBucketMetadata.java:211]: Brain fart, Pavlovian response IntelliJ squiggly line at the bottom 🐕

**Pull request merged**